protocols/mp-ecdsa/src/protocols/managers/keygen.rs

```diff
@@ -0,0 +1,554 @@
+            // for session 1
+            return self
+                .maybe_start_keygen_for_stage(
+                    KeygenRound::Next,
+                    header,
+                    anticipated_execution,
+                )
+                .await;
+        }
+
+        if matches!(state, KeygenState::Failed { session_id: 1 }) {
+            // If we are at genesis, and we have failed keygen for session 1, we may need to begin a
+            // keygen for session 1
+            return self
+                .maybe_start_keygen_for_stage(
+                    KeygenRound::Next,
+                    header,
+                    anticipated_execution,
+                )
+                .await;
+        }
+
+        None
+    }
+
+    /// Check to see if we need to run a keygen for session 2, 3, 4, .., etc.
+    async fn next_checks(
+        &self,
+        session_id: u64,
+        state: KeygenState,
+        header: &B::Header,
+        anticipated_execution: &AnticipatedKeygenExecutionStatus,
+    ) -> Option<Job> {
+        // Check bad states. These should never happen in a well-behaved program
+        if state == KeygenState::RunningGenesisKeygen {
+            self.logger.error(format!("Invalid keygen manager state: {session_id} > GENESIS_AUTHORITY_SET_ID && {state:?} == KeygenState::GenesisKeygenCompleted || {state:?} == KeygenState::RunningGenesisKeygen"));
+            return None;
+        }
+
+        if state == KeygenState::Uninitialized {
+            // We joined the network after genesis. We need to start a keygen for session `now`,
+            // so long as the next pub key isn't already on chain
+            if self.client.get_next_dkg_pub_key(header).await.is_none() {
+                return self.maybe_start_keygen_for_stage(
+                    KeygenRound::Next,
+                    header,
+                    anticipated_execution,
+                )
+                .await;
+            }
+        }
+
+        if state == KeygenState::RunningKeygen {
+            // We are in the middle of a keygen. Do nothing
+            return None;
+        }
+
+        if matches!(
+            state,
+            KeygenState::KeygenCompleted { .. } | KeygenState::Failed { .. }
+        ) {
+            // We maybe need to start a keygen for session `session_id`:
+            return self
+                .maybe_start_keygen_for_stage(
+                    KeygenRound::Next,
+                    header,
+                    anticipated_execution,
+                )
+                .await;
+        }
+
+        None
+    }
+
+    async fn maybe_start_keygen_for_stage(
+        &self,
+        stage: KeygenRound,
+        header: &B::Header,
+        anticipated_execution_status: &AnticipatedKeygenExecutionStatus,
+    ) -> Option<Job> {
+        let authority_set = if let Some((active, queued)) = self.client.validator_set(header).await {
```

Comment:
Not how it'll go. This is leading me to believe you're not processing everything about the new architecture properly given all our previous discussions.